plugins/extract/extract.plugin.zsh

```diff
@@ -4,68 +4,93 @@ extract() {
 	local remove_archive
 	local success
 	local extract_dir
+	local files
+	local unrar
+	local _7za
+	local unzip
 
 	if (( $# == 0 )); then
 		cat <<-'EOF' >&2
 			Usage: extract [-option] [file ...]
 
 			Options:
-			    -r, --remove    Remove archive after unpacking.
+			    -r, --remove                Remove archive after unpacking.
+			    -p, --password {password}   Use the password to extract.
 		EOF
 	fi
 
 	remove_archive=1
-	if [[ "$1" == "-r" ]] || [[ "$1" == "--remove" ]]; then
-		remove_archive=0
-		shift
-	fi
+	unrar="unrar x -ad"
+	unzip="unzip"
+	_7za="7za x"
+	
+	files=()
 
 	while (( $# > 0 )); do
+		if [[ "$1" == "-r" ]] || [[ "$1" == "--remove" ]]; then
+			remove_archive=0
+			shift
+			continue
+		fi
+
+		if [[ "$1" == "-p" ]] || [[ "$1" == "--password" ]]; then
+			unrar="unrar x -ad \"-p$2\""
+			_7za="7za x -p\"$2\""
+			unzip="unzip -P \"$2\""
+			shift 2
+			continue
+		fi
+
 		if [[ ! -f "$1" ]]; then
 			echo "extract: '$1' is not a valid file" >&2
 			shift
 			continue
 		fi
+		files+="$1"
+		shift
+	done
+
+	
 
+	for file in ${files[*]}; do
```

Comment:
use `"${files[@]}"` here instead of `${files[*}` to expand all elements as separate elements to iterate over.